File: android/app/src/main/java/app/noscroll/shield/SleepSchedule.kt
package app.noscroll.shield

import java.util.Calendar
import java.util.TimeZone

/**
 * Scheduled blocking (bedtime). Kotlin twin of ios/NoScrollCore SleepSchedule.swift —
 * same semantics, same edge cases, same tests.
 *
 * The bug this exists to not ship: SocialLite has two independent public reviews
 * reporting Sleep Mode failing to release at the scheduled end, one user stuck on
 * the "you should be asleep" screen at 9am. The cause is almost certainly a
 * window computed once and cached, which drifts on timezone change, DST, device
 * sleep across the boundary, and any schedule crossing midnight.
 *
 * The rule: NEVER cache the decision. [isActive] is pure and is recomputed from
 * the wall clock and the CURRENT timezone every time.
 */
data class SleepSchedule(
    /** Minutes from local midnight. */
    val startMinute: Int,
    val endMinute: Int,
    val enabled: Boolean,
    /** Calendar weekday numbers (1 = Sunday). Empty = every day. */
    val weekdays: Set<Int> = emptySet(),
) {

    fun isActive(atMillis: Long, timeZone: TimeZone = TimeZone.getDefault()): Boolean {
        if (!enabled) return false
        if (startMinute == endMinute) return false

        val cal = Calendar.getInstance(timeZone).apply { timeInMillis = atMillis }
        val nowMinute = cal.get(Calendar.HOUR_OF_DAY) * 60 + cal.get(Calendar.MINUTE)
        val weekday = cal.get(Calendar.DAY_OF_WEEK)

        val crossesMidnight = startMinute > endMinute
        val inWindow: Boolean
        val effectiveWeekday: Int

        if (crossesMidnight) {
            when {
                nowMinute >= startMinute -> {
                    inWindow = true
                    effectiveWeekday = weekday
                }
                nowMinute < endMinute -> {
                    // The morning half belongs to YESTERDAY's schedule. Getting
                    // this wrong makes a Friday-night rule fail at 1am Saturday.
                    inWindow = true
                    effectiveWeekday = if (weekday == Calendar.SUNDAY) Calendar.SATURDAY else weekday - 1
                }
                else -> {
                    inWindow = false
                    effectiveWeekday = weekday
                }
            }
        } else {
            inWindow = nowMinute in startMinute until endMinute
            effectiveWeekday = weekday
        }

        if (!inWindow) return false
        return weekdays.isEmpty() || effectiveWeekday in weekdays
    }

    companion object {
        val DEFAULT = SleepSchedule(
            startMinute = 22 * 60,
            endMinute = 7 * 60,
            enabled = false,
        )
    }
}

File: ios/NoScrollCore/Sources/NoScrollCore/SleepSchedule.swift
import Foundation

/// Scheduled blocking (bedtime).
///
/// This is the single most-reported functional bug in the app we are cloning:
/// two independent reviewers report Sleep Mode failing to release at the
/// scheduled end time — one of them stuck on the "you should be asleep" screen
/// at 9am.
///
/// The cause is almost certainly a window computed once and cached. That drifts
/// on: timezone change, DST transition, device sleep across the boundary, and
/// any schedule that crosses midnight. So this type owns exactly one rule:
///
///     NEVER cache the decision. Recompute from the wall clock and the CURRENT
///     timezone on every foreground and on a periodic tick.
///
/// `isActive(at:)` is pure, which is what makes the cases below testable without
/// a device and without waiting until 3am.
public struct SleepSchedule: Codable, Equatable, Sendable {
    /// Minutes from local midnight.
    public var startMinute: Int
    public var endMinute: Int
    public var enabled: Bool
    /// Weekday numbers (1 = Sunday, matching Calendar). Empty = every day.
    public var weekdays: Set<Int>

    public init(startMinute: Int, endMinute: Int, enabled: Bool, weekdays: Set<Int>) {
        self.startMinute = startMinute
        self.endMinute = endMinute
        self.enabled = enabled
        self.weekdays = weekdays
    }

    public static let `default` = SleepSchedule(
        startMinute: 22 * 60, endMinute: 7 * 60, enabled: false, weekdays: [])

    /// True if `date`, interpreted in `calendar`'s timezone, falls inside the window.
    ///
    /// Correctly handles a window that crosses midnight (22:00 → 07:00), which
    /// is the common case and the one a naive `start <= now && now < end` gets
    /// wrong for every minute of the night.
    public func isActive(at date: Date, calendar: Calendar = .current) -> Bool {
        guard enabled else { return false }

        let comps = calendar.dateComponents([.hour, .minute, .weekday], from: date)
        guard let hour = comps.hour, let minute = comps.minute, let weekday = comps.weekday else {
            return false
        }
        let nowMinute = hour * 60 + minute

        if startMinute == endMinute { return false }

        let crossesMidnight = startMinute > endMinute
        let inWindow: Bool
        let effectiveWeekday: Int

        if crossesMidnight {
            if nowMinute >= startMinute {
                // Evening portion — belongs to today.
                inWindow = true
                effectiveWeekday = weekday
            } else if nowMinute < endMinute {
                // Morning portion — belongs to YESTERDAY's schedule. Getting
                // this wrong is how a Friday-night schedule fails to cover
                // Saturday 1am.
                inWindow = true
                effectiveWeekday = weekday == 1 ? 7 : weekday - 1
            } else {
                inWindow = false
                effectiveWeekday = weekday
            }
        } else {
            inWindow = nowMinute >= startMinute && nowMinute < endMinute
            effectiveWeekday = weekday
        }

        guard inWindow else { return false }
        return weekdays.isEmpty || weekdays.contains(effectiveWeekday)
    }
}
